language/move-model/src/builder/exp_translator.rs

```diff
@@ -1785,6 +1788,94 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
     }
 
+    fn translate_quant(
+        &mut self,
+        loc: &Loc,
+        kind: PA::QuantKind,
+        ranges: &EA::LValueWithRangeList,
+        body: &EA::Exp,
+        expected_type: &Type,
+    ) -> Exp {
+        // Enter the quantifier variables into a new local scope and collect their declarations.
+        self.enter_scope();
+        let mut rranges = vec![];
+        for range in &ranges.value {
+            // The quantified variable and its domain expression.
+            let (bind, exp) = &range.value;
+            let loc = self.to_loc(&bind.loc);
+            let (exp_ty, rexp) = self.translate_exp_free(exp);
+            let ty = self.fresh_type_var();
+            let exp_ty = self.subs.specialize(&exp_ty);
+            match &exp_ty {
+                Type::Vector(..) => {
+                    self.check_type(
+                        &loc,
+                        &exp_ty,
+                        &Type::Vector(Box::new(ty.clone())),
+                        "in quantification over vector",
+                    );
+                }
+                Type::TypeDomain(..) => {
+                    self.check_type(
+                        &loc,
+                        &exp_ty,
+                        &Type::TypeDomain(Box::new(ty.clone())),
+                        "in quantification over domain",
+                    );
+                }
+                Type::Primitive(PrimitiveType::Range) => {
+                    self.check_type(
+                        &loc,
+                        &ty,
+                        &Type::Primitive(PrimitiveType::Num),
+                        "in quantification over range",
+                    );
+                }
+                _ => {
+                    self.error(&loc, "quantified variables must range over a vector, a type domain, or a number range");
+                    return self.new_error_exp();
+                }
+            }
+            match &bind.value {
+                EA::LValue_::Var(
+                    Spanned {
+                        value: EA::ModuleAccess_::Name(n),
+                        ..
+                    },
+                    _,
+                ) => {
+                    let name = self.symbol_pool().make(&n.value);
+                    let id = self.new_node_id_with_type_loc(&ty, &loc);
+                    self.define_local(&loc, name, ty.clone(), None);
+                    let rbind = LocalVarDecl {
+                        id,
+                        name,
+                        binding: None,
+                    };
+                    rranges.push((rbind, rexp));
+                }
+                EA::LValue_::Unpack(..) | EA::LValue_::Var(..) => self.error(
+                    &loc,
+                    "[current restriction] tuples not supported in quantifiers",
+                ),
+            }
+        }
+        let rty = self.check_type(
+            loc,
+            &Type::new_prim(PrimitiveType::Bool),
```

Comment:
nit: there is a constant BOOL_TYPE in type.rs

Reply:
It looks like this constant is (uniformly) not used in this file yet.